File: skills/kinetica-execute/scripts/kinetica-cli.py
#!/usr/bin/env python3
"""Kinetica GPU Database CLI -- thin dispatcher.

Routes commands to core or category modules while maintaining backward
compatibility with all 10 original flat commands.
"""

import warnings

# Suppress urllib3 warning about LibreSSL on macOS system Python.
# gpudb requires urllib3>=2 which expects OpenSSL 1.1.1+, but macOS Xcode
# Command Line Tools ship LibreSSL. The warning is cosmetic -- requests work fine.
warnings.filterwarnings("ignore", message=r"urllib3 v2.*only supports OpenSSL")

import importlib
import sys
from pathlib import Path

# Ensure the scripts directory is on sys.path so ``from modules.xxx`` works
_SCRIPTS_DIR = str(Path(__file__).resolve().parent)
if _SCRIPTS_DIR not in sys.path:
    sys.path.insert(0, _SCRIPTS_DIR)

from modules.helpers import connect, die, env, load_env_file, out  # noqa: E402
from modules.core import COMMANDS as CORE_COMMANDS, build_parser  # noqa: E402

# Ensure .env is loaded (helpers auto-loads on import, but be explicit)
load_env_file()

# ---------------------------------------------------------------------------
# Category modules
# ---------------------------------------------------------------------------

CATEGORY_MODULES = {
    "graph": "modules.graph",
    "geo": "modules.geo",
    "viz": "modules.viz",
    "io": "modules.io_cmd",
    "monitor": "modules.monitor",
}


def _print_help():
    """Print combined help covering core commands and available categories."""
    parser = build_parser()
    parser.print_help()

    # List categories that are actually importable
    available = []
    for name, mod_path in sorted(CATEGORY_MODULES.items()):
        try:
            mod = importlib.import_module(mod_path)
            actions = ", ".join(sorted(getattr(mod, "COMMANDS", {}).keys()))
            available.append(f"  {name:18s} {actions or '(no actions)'}")
        except ImportError:
            pass

    if available:
        print("\nCategories:")
        for line in available:
            print(line)


def _run_handler(fn, db, args):
    """Run a command function with shared error handling."""
    try:
        fn(db, args)
    except Exception as e:
        msg = str(e)
        if "Unable to sort on array column" in msg:
            out({"error": msg, "fix": 'Remove the array column from ORDER BY, use a non-array column, or index into it: ORDER BY "col"[1]'})
        else:
            out({"error": msg})
        sys.exit(1)


def _dispatch_category(category, argv_rest):
    """Load a category module and dispatch to the requested action."""
    mod_path = CATEGORY_MODULES[category]
    try:
        mod = importlib.import_module(mod_path)
    except ImportError as exc:
        die(f"Category '{category}' is not available: {exc}")

    commands = getattr(mod, "COMMANDS", {})
    if not argv_rest or argv_rest[0] in ("--help", "-h"):
        lines = [f"Usage: kinetica-cli {category} <action> [args]\n", "Actions:"]
        for action_name, info in sorted(commands.items()):
            lines.append(f"  {action_name:18s} {info.get('desc', '')}")
        print("\n".join(lines))
        sys.exit(0)

    action = argv_rest[0]
    entry = commands.get(action)
    if not entry:
        die(f"Unknown action '{action}' in category '{category}'. "
            f"Available: {', '.join(sorted(commands.keys()))}")

    # Build an argparse parser for this action and parse remaining args
    import argparse
    action_parser = argparse.ArgumentParser(
        prog=f"kinetica-cli {category} {action}",
        description=entry.get("desc", ""),
    )
    build_args_fn = entry.get("build_args")
    if build_args_fn:
        build_args_fn(action_parser)

    args = action_parser.parse_args(argv_rest[1:])

    db = _connect_or_die()
    _run_handler(entry["fn"], db, args)


def _connect_or_die():
    """Connect to Kinetica, handling common connection errors."""
    try:
        return connect()
    except Exception as e:
        err_str = str(e)
        if "Connection refused" in err_str or "Failed to get database version" in err_str:
            url = env("KINETICA_DB_SKILL_URL", "(not set)")
            out({"error": f"Cannot connect to Kinetica at {url}. "
                          "Check KINETICA_DB_SKILL_URL and ensure the server is running."})
        else:
            out({"error": err_str[:500]})
        sys.exit(1)


def main():
    argv = sys.argv[1:]

    # No args or explicit help
    if not argv or argv[0] in ("--help", "-h"):
        _print_help()
        sys.exit(0)

    first = argv[0]

    # Category dispatch
    if first in CATEGORY_MODULES:
        _dispatch_category(first, argv[1:])
        return

    # Core command dispatch via argparse
    parser = build_parser()
    args = parser.parse_args(argv)

    if not args.command:
        parser.print_help()
        sys.exit(0)

    fn = CORE_COMMANDS.get(args.command)
    if not fn:
        die(f"Unknown command: {args.command}")

    db = _connect_or_die()
    _run_handler(fn, db, args)


if __name__ == "__main__":
    main()

File: plugins/kinetica/skills/kinetica-execute/scripts/kinetica-cli.js
#!/usr/bin/env node
'use strict';

/**
 * Kinetica GPU Database CLI -- thin dispatcher.
 *
 * Routes commands to core or category modules while maintaining backward
 * compatibility with all 10 original flat commands.
 */

const { connect, parseArgs, die, out } = require('./modules/helpers');

const CORE_COMMANDS = require('./modules/core');

// ---------------------------------------------------------------------------
// Category modules
// ---------------------------------------------------------------------------

const CATEGORY_MODULES = {
  graph: './modules/graph',
  geo: './modules/geo',
  viz: './modules/viz',
  io: './modules/io_cmd',
  monitor: './modules/monitor',
};

// ---------------------------------------------------------------------------
// Help
// ---------------------------------------------------------------------------

function printHelp() {
  const lines = ['Usage: kinetica-cli <command> [args]\n', 'Commands:'];
  for (const [name, { desc }] of Object.entries(CORE_COMMANDS)) {
    lines.push(`  ${name.padEnd(18)} ${desc}`);
  }

  // List categories that are actually importable
  const categories = [];
  for (const [name, modPath] of Object.entries(CATEGORY_MODULES)) {
    try {
      const mod = require(modPath);
      const actions = Object.keys(mod).sort().join(', ');
      categories.push(`  ${name.padEnd(18)} ${actions || '(no actions)'}`);
    } catch (_) {
      // Module not yet implemented -- skip
    }
  }

  if (categories.length > 0) {
    lines.push('\nCategories:');
    for (const line of categories) {
      lines.push(line);
    }
  }

  lines.push(
    '\nEnvironment variables: KINETICA_DB_SKILL_URL, KINETICA_DB_SKILL_USER, KINETICA_DB_SKILL_PASS, KINETICA_DB_SKILL_OAUTH_TOKEN, KINETICA_DB_SKILL_TIMEOUT'
  );
  console.log(lines.join('\n'));
}

// ---------------------------------------------------------------------------
// Shared error handler for command dispatch
// ---------------------------------------------------------------------------

async function runHandler(fn, db, args) {
  try {
    await fn(db, args);
  } catch (err) {
    const msg = err.message || String(err);
    if (msg.includes('Unable to sort on array column')) {
      out({
        error: msg,
        fix: 'Remove the array column from ORDER BY, use a non-array column, or index into it: ORDER BY "col"[1]',
      });
    } else {
      out({ error: msg });
    }
    process.exit(1);
  }
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
  const argv = process.argv.slice(2);

  if (argv.length === 0 || argv[0] === '--help' || argv[0] === '-h') {
    printHelp();
    process.exit(0);
  }

  const first = argv[0];

  // Category dispatch
  if (first in CATEGORY_MODULES) {
    const modPath = CATEGORY_MODULES[first];
    let mod;
    try {
      mod = require(modPath);
    } catch (err) {
      die(`Category '${first}' is not available: ${err.message}`);
    }

    const rest = argv.slice(1);
    if (rest.length === 0 || rest[0] === '--help' || rest[0] === '-h') {
      const actionLines = [`Usage: kinetica-cli ${first} <action> [args]\n`, 'Actions:'];
      for (const [actionName, info] of Object.entries(mod)) {
        actionLines.push(`  ${actionName.padEnd(18)} ${info.desc || ''}`);
      }
      console.log(actionLines.join('\n'));
      process.exit(0);
    }

    const action = rest[0];
    const entry = mod[action];
    if (!entry) {
      die(
        `Unknown action '${action}' in category '${first}'. ` +
          `Available: ${Object.keys(mod).sort().join(', ')}`
      );
    }

    const args = parseArgs(rest);
    const db = connect();
    await runHandler(entry.fn, db, args);
    return;
  }

  // Core command dispatch
  const args = parseArgs(argv);
  const cmd = CORE_COMMANDS[args.cmd];
  if (!cmd) die(`Unknown command: ${args.cmd}. Run with --help for usage.`);

  const db = connect();
  await runHandler(cmd.fn, db, args);
}

// Auto-run when invoked directly; skip when require()'d by tests
if (require.main === module) {
  main();
}

module.exports = { main };
